lmnet/lmnet/datasets/tensorflow_datasets.py

```diff
@@ -0,0 +1,141 @@
+class TensorFlowDatasetsBase(Base):
+    """
+    Abstract dataset class for loading TensorFlow Datasets.
+    The parameter "extend_dir" should be a path to the prepared data of tfds.
+    "extend_dir" can be a path either in local or in Google Cloud Storage.
+    """
+    available_subsets = ["train", "train_validation_saving", "validation"]
+
+    def __init__(
+            self,
+            *args,
+            **kwargs
+    ):
+        super().__init__(
+            *args,
+            **kwargs,
+        )
+
+        # The last directory name of extend_dir is the dataset name
+        splitted_path = self.extend_dir.rstrip('/').split('/')
+        data_dir = '/'.join(splitted_path[:-1])
+        name = splitted_path[-1]
+        builder = tfds.builder(name, data_dir=data_dir)
+
+        self.info = builder.info
+        self._init_available_splits()
+
+        self.dataset = builder.as_dataset(split=self.available_splits[self.subset])
+        self._init_features()
+
+    def _init_available_splits(self):
+        """
+        Initializing available splits dictionary depending on
+        what kind of splits the dataset has.
+        """
+        self.available_splits = {}
+        if tfds.Split.TRAIN not in self.info.splits:
+            raise ValueError("Datasets need to have a split \"TRAIN\".")
+
+        if tfds.Split.VALIDATION in self.info.splits and tfds.Split.TEST in self.info.splits:
+            self.available_splits["train"] = tfds.Split.TRAIN
+            self.available_splits["validation"] = tfds.Split.VALIDATION
+            self.available_splits["train_validation_saving"] = tfds.Split.TEST
+
+        elif tfds.Split.VALIDATION in self.info.splits:
+            self.available_splits["train"] = tfds.Split.TRAIN
+            self.available_splits["validation"] = tfds.Split.VALIDATION
+
+        elif tfds.Split.TEST in self.info.splits:
+            self.available_splits["train"] = tfds.Split.TRAIN
+            self.available_splits["validation"] = tfds.Split.TEST
+
+        else:
+            train, validation = tfds.Split.TRAIN.subsplit([9, 1])
+            self.available_splits["train"] = train
+            self.available_splits["validation"] = validation
+
+    def _init_features(self):
+        """
+        Initializing feature variables by using tf.Tensor from datasets.
+        For classification, "images" and "labels" should be initialized.
+        For object detection, "images" and "gt_boxes" should be initialized.
+        """
+        raise NotImplementedError()
+
+    @property
+    def num_per_epoch(self):
+        return len(self.images)
+
+
+class TensorFlowDatasetsClassification(TensorFlowDatasetsBase):
+    """
+    A dataset class for loading TensorFlow Datasets for classification.
+    """
+    @property
+    def classes(self):
+        return self.info.features["label"].names
+
+    @property
+    def num_classes(self):
+        return self.info.features["label"].num_classes
+
+    def _init_features(self):
```

Comment:
Do `self.images` and `self.labels` have all of data on memory?
It is better to get a data for each `__getitem__` for big dataset.

Reply:
Yes, they have all the data on memory. I agree that it's not good for big datasets but there is no way to access TensorFlow Datasets by index because TensorFlow Datasets work as iterators. So current implementation is a workaround. To solve this problem, we need to implement this class as a new dataset iterator.

Reply:
Another idea is tha we can ignore index in the ```__getitem__``` method.

e.g. load 10000 data into memory (from TF datasets iterator), shuffle, then always pop the 1st item in the list in ```__getitem__``` method.

That's not a straightforward implementation, but as a tentative solution, we can tolerate it.

Reply:
@tkng
Thanks for your suggestion. I think it's a good idea.
Currently, the method `__getitem__` seems to be used only for shuffling in dataset_iterator.

Reply:
I tried to update `__getitem__` as we talked above. I used `tf.Session().run()` for getting next element but I found that it doesn't work with the option `enable_prefetch = True`. Maybe we can fix this problem by changing `DatasetIterator` in order to use tf.Session().run() in multi processes safely.

https://github.com/blue-oil/blueoil/blob/cf5c085933aba3b7d7f319468518fd557b1322ff/lmnet/lmnet/datasets/tensorflow_datasets.py#L62-L67

But I think we should not change many things only for introducing TensorFlowDatasets class because this class is a temporary solution anyway. I and @ruimashita talked and decided not to solve the problem this time. All the data will be loaded on initialization of the dataset class.